controlplane/rosa/controllers/rosacontrolplane_controller.go

```diff
@@ -352,6 +365,115 @@ func (r *ROSAControlPlaneReconciler) reconcileDelete(ctx context.Context, rosaSc
+	apiServerUrl := cluster.API().URL()
+
+	err = rosaClient.CreateAdminUserIfNotExist(cluster.ID(), userName, password)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	clientConfig := &restclient.Config{
+		Host:     apiServerUrl,
+		Username: userName,
+	}
+	token, err := rosa.RequestToken(apiServerUrl, userName, password, clientConfig)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to request token: %w", err)
+	}
+
+	contextName := fmt.Sprintf("%s@%s", userName, clusterName)
+	cfg := &api.Config{
+		APIVersion: api.SchemeGroupVersion.Version,
+		Clusters: map[string]*api.Cluster{
+			clusterName: {
+				Server: apiServerUrl,
+			},
+		},
+		Contexts: map[string]*api.Context{
+			contextName: {
+				Cluster:  clusterName,
+				AuthInfo: userName,
+			},
+		},
+		CurrentContext: contextName,
+		AuthInfos: map[string]*api.AuthInfo{
+			userName: {
+				Token: token.AccessToken,
+			},
+		},
+	}
+	out, err := clientcmd.Write(*cfg)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to serialize config to yaml: %w", err)
+	}
+
+	if kubeconfigSecret != nil {
+		// update existing kubeconfig secret
+		kubeconfigSecret.Data[secret.KubeconfigDataName] = out
+		if err := r.Client.Update(ctx, kubeconfigSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update kubeconfig secret: %w", err)
+		}
+	} else {
+		// create new kubeconfig secret
+		controllerOwnerRef := *metav1.NewControllerRef(rosaScope.ControlPlane, rosacontrolplanev1.GroupVersion.WithKind("ROSAControlPlane"))
+		kubeconfigSecret = kubeconfig.GenerateSecretWithOwner(clusterRef, out, controllerOwnerRef)
+		if err := r.Client.Create(ctx, kubeconfigSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create kubeconfig secret: %w", err)
+		}
+	}
+
+	rosaScope.ControlPlane.Status.Initialized = true
+	// requeue to refresh access token when it expires.
+	return ctrl.Result{RequeueAfter: token.ExpiresIn}, nil
+}
+
+func (r *ROSAControlPlaneReconciler) reconcileClusterAdminPassword(ctx context.Context, rosaScope *scope.ROSAControlPlaneScope) (string, error) {
+	passwordSecret := rosaScope.ClusterAdminPasswordSecret()
+	err := r.Client.Get(ctx, client.ObjectKeyFromObject(passwordSecret), passwordSecret)
+	if err == nil {
+		password := string(passwordSecret.Data["value"])
+		return password, nil
+	} else if !apierrors.IsNotFound(err) {
+		return "", fmt.Errorf("failed to get cluster admin password secret: %w", err)
+	}
+	// Generate a new password and create the secret
+	password, err := rosa.GenerateRandomPassword()
+	if err != nil {
+		return "", err
+	}
+
+	controllerOwnerRef := *metav1.NewControllerRef(rosaScope.ControlPlane, rosacontrolplanev1.GroupVersion.WithKind("ROSAControlPlane"))
+	passwordSecret.Data = map[string][]byte{
+		"value": []byte(password),
+	}
+	passwordSecret.OwnerReferences = []metav1.OwnerReference{
+		controllerOwnerRef,
+	}
+	if err := r.Client.Create(ctx, passwordSecret); err != nil {
+		return "", err
+	}
```

Comment:
What's this used for?

Reply:
After the user is created, there is no way to retrieve the password from the ocm API. So if we failed to request the token or create the kubeconfig, we wouldn't be able to get the password in the next reconciliation loop unless we stored it.
Actually, the first couple of token requests always fail, because the user has not been fully registered yet.

Reply:
sounds good, let's add a comment to that effect?

Reply:
done